test/tests/functional/pbs_job_array.py

```diff
@@ -744,3 +782,237 @@ def test_recover_big_array_job(self):
+        j2_id = self.server.submit(j2)
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=j2_id)
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+        j2_sub1 = j2.create_subjob_id(j2_id, 1)
+        job_arr = self.server.status(JOB, id=j2_sub1)
+        stime = self.lu.convert_date_time(job_arr[0]['stime'],
+                                          fmt="%a %b %d %H:%M:%S %Y")
+        job_arr = self.server.status(JOB, id=j2_id)
+
+        # check estimated start time is set on job array
+        self.assertIn('estimated.start_time', job_arr[0])
+        errmsg = j2_id + ";Error in calculation of start time of top job"
+        self.scheduler.log_match(errmsg, existence=False, max_attempts=10)
+        est = self.lu.convert_date_time(job_arr[0]['estimated.start_time'],
+                                        fmt="%a %b %d %H:%M:%S %Y")
+        self.assertAlmostEqual(stime + 300, est, 1)
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_queuejob_hook(self):
+        """
+        Test that a queuejob hook is able to set max_run_subjobs attribute.
+        """
+        a = {'resources_available.ncpus': 8}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        self.create_max_run_subjobs_hook(3, "queuejob", "h1")
+        j1 = Job(attrs={ATTR_J: '1-20'})
+        jid1 = self.server.submit(j1)
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.expect(JOB, {ATTR_state: 'B'}, id=jid1)
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+        # Submit a normal job and see if queuejob hook cannot set the
+        # attribute.
+        with self.assertRaises(PbsSubmitError) as e:
+            self.server.submit(Job())
+        self.assertIn("Not an array job", e.exception.msg[0])
+
+    @skipOnCpuSet
+    def test_max_run_subjobs_modifyjob_hook(self):
+        """
+        Submit array job with large max_run_subjobs limit see if modifyjob
+        modifies it.
+        """
+        a = {'resources_available.ncpus': 20}
+        self.server.manager(MGR_CMD_SET, NODE, a, self.mom.shortname)
+
+        self.create_max_run_subjobs_hook(3, "modifyjob", "h1")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+        j = Job(attrs={ATTR_J: '1-50'})
+        jid = self.server.submit(j)
+        self.server.expect(JOB, {ATTR_state: 'Q'}, id=jid)
+        self.server.alterjob(jid, {ATTR_W: 'max_run_subjobs=20'})
+        self.server.log_match("max_run_subjobs set to 3")
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'True'})
+        self.server.expect(JOB, {'job_state=R': 3}, extend='t')
+
+        # Modify a normal job and see if queuejob hook cannot set the
+        # attribute.
+        self.server.manager(MGR_CMD_SET, SERVER, {'scheduling': 'False'})
+        nj = self.server.submit(Job())
+        with self.assertRaises(PbsAlterError) as e:
+            self.server.alterjob(nj, {ATTR_W: 'max_run_subjobs=20'})
```

Comment:
Can you try this with altering ncpus or something and see if the hook rejects?

Reply:
I can do that, however it will require me to write a new modify job hook because the one I have does not set this attribute if it is already set to 0 (which when modifying ncpus will be the case). I'll add that to the test.